### PR TITLE
detect/xor: implement xor keywords

### PR DESCRIPTION
**Pull request by simdugas**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
- [#3285](https://redmine.openinfosecfoundation.org/issues/3285)

Link to suricata-verify PR:
- [#243](https://github.com/OISF/suricata-verify/pull/243)

Describe changes:
- Implements `xor` and `xor_data` keywords to decrypt buffer data and place it into a sticky buffer for further inspection.
- Added utility functions to decode hex data and to decrypt data using a simple xor cypher.
- More details about these keywords can be found in the updated user documentation at `doc/userguide/rules/xor-keywords.rst`. 

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output (if applicable):
- passes on local docker buildbot without the `--enable-prelude` flag. There seems to be an unrelated build failure in the latest master:
```
alert-prelude.c: In function ‘PacketToDataProtoEmail’:
alert-prelude.c:784:10: error: implicit declaration of function ‘JsonEmailAddMetadata’; did you mean ‘EveEmailAddMetadata’? [-Werror=implicit-function-declaration]
     js = JsonEmailAddMetadata(p->flow, pa->tx_id);
          ^~~~~~~~~~~~~~~~~~~~
          EveEmailAddMetadata
```

### CODE REVIEW

File: doc/userguide/rules/xor-keywords.rst [hunk 1]
```diff
@@ -0,0 +1,118 @@
+XOR keywords
+===============
```
Comment (norg) [inline doc/userguide/rules/xor-keywords.rst:2]: align the amount of `=` to the headline, same through the rest of the document.

File: src/detect-xor.c [hunk 2]
```diff
@@ -0,0 +1,1080 @@
...
+
+    /* copy the key when using a byte_extract variable */
+    if (data->bid >= 0) {
+        uint64_t value = det_ctx->bj_values[data->bid];
+
+        /* DetectXorParse() will ensure that data->key_len is between 1 and 8
+         * (inclusive) when a byte_extract variable is used. */
+        for (int i = 0; i < (int) data->key_len; i++) {
+            /* Do the opposite of ByteExtract() assuming big endianness. */
+            data->key[data->key_len - i - 1] = value >> ((i & 7) << 3);
```
[the remark this answers is not in the document]
Comment (victorjulien) [reply src/detect-xor.c:101]: I can test on my ppc64 box.

File: src/detect-xor.c [hunk 3]
```diff
@@ -0,0 +1,1080 @@
...
+                        key_str, data->key_len);
+                goto cleanup;
+            }
+
+            /** \todo improve cache coherence by embedding the key array inside
+             * the context struct instead of allocating it. This could work well
+             * for short key sizes (i.e. up to 8 bytes) and be represented as a
+             * union. This would require us to track when we choose to allocate
+             * using a flag. */
+            data->key = SCCalloc(data->key_len, sizeof(*data->key));
```
[the remark this answers is not in the document]
Comment (victorjulien) [reply src/detect-xor.c:243]: Lets first do some testing before adding this.

File: src/detect-xor.c [hunk 4]
```diff
@@ -0,0 +1,1080 @@
...
+    if (data->key == NULL && data->bid < 0) {
+        SCLogError(SC_ERR_INVALID_RULE_ARGUMENT,
+                "Bad value for xor key: \"%s\". "
+                "Expected hex string in double quotes or byte_extract variable.",
+                key_str);
+        goto cleanup;
+    }
+
+     if (pcre_rc >= 3) {
+        if (pcre_get_substring((char *)str, ov, max, 3, &bytes_str)) {
```
Comment (victorjulien) [inline src/detect-xor.c:260]: in general we strive to use pcre_copy_substring over pcre_get_substring as the latter allocs on the heap.

**Comment by simdugas** [comment]
> Thanks! It's clear that a lot of work has gone into this.
> 
> Some minor comments from a quick glance.
> 
> More fundamentally I wonder if this should be implemented as a 'transform' instead, where we'd do something like:
> 
> `http.uri; xor:<key>[,<other opts>]; content:"<match>";`
> 
> This would work similar to how `to_md5`, 'strip_whitespace` and friends modify the buffer in-place. Options for transforms are coming soon.

Thanks for the feedback, I will look into implementing this as a transform.

My first approach was to implement it as an inspect engine but I couldn't get it to match on the payload itself. I then based it off of the `base64` implementation, but used an `InspectionBuffer` to store the decrypted contents instead of adding more fields to the `de_ctx`.

**Pull request closed without merging**